setup.py

```diff
@@ -14,172 +14,91 @@
-    }
 
-    l_opts = {
-        "msvc": ["/WX"],
-        "unix": ["-O3", "-fPIC", "-shared", "-fopenmp", "-Wall", "-Wextra", "-Werror"],
-    }
-    if platform.system() == "Linux" and platform.machine() == "x86_64":
-        c_opts["unix"].append("-mavx")
-        l_opts["unix"].append("-mavx")
-
-    if platform.system() == "Darwin":
-        for opts in (c_opts, l_opts):
-            opts["unix"].remove("-fopenmp")
-            opts["unix"].remove("-shared")
-
-        darwin_opts = ["-stdlib=libc++"]
-
-        # Used to enable OpenMP only on Intel 
-        # Macs due to CI available of M1
-        if os.environ.get("USE_OMP"):
-            darwin_opts.extend([
-                "-Xpreprocessor",
-                "-fopenmp",
-            ])
+    user_options = build_ext.user_options + [
+        ("define=", "D", "Define variables for CMake")
+    ]
 
-        darwin_opts.append("-mmacosx-version-min=10.14")
+    def initialize_options(self):
+        super().initialize_options()
+        self.define = None
+
+    def finalize_options(self):
+        # Parse the custom CMake options and store them in a new attribute
+        defines = [] if self.define is None else self.define.split(";")
+        self.cmake_defines = [f"-D{define}" for define in defines]
+
+        super().finalize_options()
+
+    def build_extension(self, ext: CMakeExtension):
+        extdir = str(Path(self.get_ext_fullpath(ext.name)).parent.absolute())
+
+        debug = int(os.environ.get("DEBUG", 0)) if self.debug is None else self.debug
+        cfg = "Debug" if debug else "Release"
+        ninja_path = str(shutil.which('ninja'))
+
+        # Set Python_EXECUTABLE instead if you use PYBIND11_FINDPYTHON
+        configure_args = [
+            "-GNinja",
+            f"-DCMAKE_LIBRARY_OUTPUT_DIRECTORY={extdir}",
+            f"-DPYTHON_EXECUTABLE={sys.executable}",
+            f"-DCMAKE_BUILD_TYPE={cfg}",  # not used on MSVC, but no harm
+            f"-DCMAKE_MAKE_PROGRAM={ninja_path}",
+        ]
+        configure_args += self.cmake_defines
         
-        c_opts["unix"] += darwin_opts
-        l_opts["unix"] += darwin_opts
-
-    def build_extensions(self):
-        ct = self.compiler.compiler_type
-        opts = self.c_opts.get(ct, [])
-        link_opts = self.l_opts.get(ct, [])
-
-        if ct == "unix":
-            opts.append(cpp_flag(self.compiler))
-            if has_flag(self.compiler, "-fvisibility=hidden"):
-                opts.append("-fvisibility=hidden")
-
-        for ext in self.extensions:
-            ext.define_macros = [("VERSION_INFO", '"{}"'.format(self.distribution.get_version()))]
-            ext.extra_compile_args = opts
-            ext.extra_link_args = link_opts
-
-        build_ext.build_extensions(self)
-
-
-if not os.environ.get("SKIP_COMPILATION", False):
+        build_args = []
+
+        if platform.system() == "Darwin":
+            configure_args += []
```

Comment:
One more change I think we need here is the OpenMP support being disabled on the M1 Macs. I had a workaround until Github Actions add M1 machines to their CI at https://github.com/PennyLaneAI/pennylane-lightning/blob/63929edacc85840b88208bfc0ef9dd75901e2cf6/setup.py#L98

Basically, if we enable the env variable the Mac build with or without OpenMP support. This comes from us using libomp in brew, which only pulls in the x86_64 version, even when building for M1 Macs (as they are still running on x86_64 in Github actions).

Reply:
Thank you for noting this. I just thought it was fine as Wheel::MacOS::ARM didn't work for PRs. Now a wheel builder for M1 also works without an error.